CoseSignTool/CoseSign1/DetachedSignatureFactory.cs

```diff
@@ -0,0 +1,301 @@
+    {
+        if (string.IsNullOrWhiteSpace(contentType))
+        {
+            throw new ArgumentNullException(nameof(contentType), "A content type must be specified");
+        }
+
+        ReadOnlyMemory<byte> hash = InternalHashAlgorithm.ComputeHash(payload.ToArray());
+        return Task.FromResult(InternalMessageFactory.CreateCoseSign1MessageBytes(
+            new MemoryStream(hash.ToArray()),
+            signingKeyProvider,
+            embedPayload: true,
+            contentType: ExtendContentType(contentType)));
+    }
+
+    /// <summary>
+    /// Creates a detached signature of the specified payload returned as a <see cref="CoseSign1Message"/> following the rules in this class description.
+    /// </summary>
+    /// <param name="payload">The payload to create a detached signature for.</param>
+    /// <param name="signingKeyProvider">The COSE signing key provider to be used for the signing operation within the <see cref="ICoseSign1MessageFactory"/>.</param>
+    /// <param name="contentType">A media type string following https://datatracker.ietf.org/doc/html/rfc6838.</param>
+    /// <returns>A byte[] representation of a COSE Sign1 Message which can be used as a detached signature validation of the payload.</returns>
+    /// <exception cref="ArgumentNullException">The contentType parameter was empty or null</exception>
+    public ReadOnlyMemory<byte> CreateDetachedSignatureBytes(
+        Stream payload,
+        ICoseSigningKeyProvider signingKeyProvider,
+        string contentType)
+    {
+        if (string.IsNullOrWhiteSpace(contentType))
+        {
+            throw new ArgumentNullException(nameof(contentType), "A content type must be specified");
+        }
+
+        ReadOnlyMemory<byte> hash = InternalHashAlgorithm.ComputeHash(payload);
+        return InternalMessageFactory.CreateCoseSign1MessageBytes(
+            new MemoryStream(hash.ToArray()),
+            signingKeyProvider,
+            embedPayload: true,
+            contentType: ExtendContentType(contentType));
+    }
+
+    /// <summary>
+    /// Creates a detached signature of the specified payload returned as a <see cref="CoseSign1Message"/> following the rules in this class description.
+    /// </summary>
+    /// <param name="payload">The payload to create a detached signature for.</param>
+    /// <param name="signingKeyProvider">The COSE signing key provider to be used for the signing operation within the <see cref="ICoseSign1MessageFactory"/>.</param>
+    /// <param name="contentType">A media type string following https://datatracker.ietf.org/doc/html/rfc6838.</param>
+    /// <returns>A Task which when completed returns a byte[] representation of a COSE Sign1 Message which can be used as a detached signature validation of the payload.</returns>
+    /// <exception cref="ArgumentNullException">The contentType parameter was empty or null</exception>
+    public Task<ReadOnlyMemory<byte>> CreateDetachedSignatureBytesAsync(

```

Comment:
I like that you have lots of overloads here for different user scenarios. Look like you settled on a matrix of input byts|stream, sync|async, and return CoseSign1Message|bytes.
I also see lots of opportunity for code re-use. Personally, I would put all the code into two private/internal methods: CreateDetachedSignatureInternal and CreateDetachedSignatureBytesInternal, and let the public methods just be overloads and task wrappers.